### PR TITLE
If limit not reached after a certain amount iterators, estimate how m…

### PR DESCRIPTION
**Pull request by baldersheim**
…any iterators are needed and

step iterator forward.

@toregge and @geirst PR
I think we need -= and step backwards too, or ability to step in negative direction.

### CODE REVIEW

File: searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp [hunk 1]
```diff
@@ -244,25 +244,46 @@ PostingListSearchContextT<DataT>::approximateHits() const
...
+            }
+        }
+
     } else if ((rangeLimit < 0) && (_lowerDictItr != _upperDictItr)) {
         rangeLimit = -rangeLimit;
         DictionaryConstIterator middle = _upperDictItr;
-        for (int n(0); (n < rangeLimit) && (middle != _lowerDictItr); ) {
+        long n = 0;
+        size_t count = 0;
+        for (; (n < rangeLimit) && (count < max_posting_lists_to_count) && (middle != _lowerDictItr); ) {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp:271]: I think `count` must be incremented as well?

File: searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp [hunk 2]
```diff
@@ -244,25 +244,46 @@ PostingListSearchContextT<DataT>::approximateHits() const
...
             --middle;
             n += _postingList.frozenSize(middle.getData().load_acquire());
         }
-        _lowerDictItr = middle;
-        _uniqueValues = _upperDictItr - _lowerDictItr;
+        if (n >= rangeLimit) {
+            _lowerDictItr = middle;
+        } else {
+            size_t offset = ((rangeLimit - n) * count)/n;
+            //TODO middle -= offset;
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp:279]: This is now supported in https://github.com/vespa-engine/vespa/pull/29405.

**Review by geirst — COMMENTED** [review]
👍 But one comment that must be addressed first.

File: searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp (hunk 2 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp]: Thanks, used now

File: searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp]: Correct, the tests told me the same.

**Comment by baldersheim** [comment]
@geirst PTAL

**Review by geirst — DISMISSED** [review]
👍 @toregge please review as well

File: build_settings.cmake [hunk 3]
```diff
@@ -97,7 +97,7 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND VESPA_USE_LTO)
 endif()
 
 # C and C++ compiler flags
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -O3 -fno-omit-frame-pointer ${C_WARN_OPTS} -fPIC ${VESPA_CXX_ABI_FLAGS} ${VESPA_XXHASH_DEFINE} -DBOOST_DISABLE_ASSERTS ${VESPA_CPU_ARCH_FLAGS} ${EXTRA_C_FLAGS}")
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -O0 -fno-omit-frame-pointer ${C_WARN_OPTS} -fPIC ${VESPA_CXX_ABI_FLAGS} ${VESPA_XXHASH_DEFINE} -DBOOST_DISABLE_ASSERTS ${VESPA_CPU_ARCH_FLAGS} ${EXTRA_C_FLAGS}")
```
Comment (geirst) [inline build_settings.cmake:100]: Was this change intentional?

Comment (baldersheim) [reply build_settings.cmake:100]: Not at all... :)

**Review by toregge — DISMISSED** [review]

**Comment by geirst** [comment]
This depends on refactor changes in https://github.com/vespa-engine/vespa/pull/29415, and will need a rebase.

**Comment by aressem** [comment]
This one might have broken the `ReportCoverage::test_reportcoverage` test
